caikit/core/task.py

```diff
@@ -0,0 +1,180 @@
+        raise NotImplementedError("This is implemented by the @task decorator!")
+
+    @classmethod
+    def get_task_group(cls) -> Type[TaskGroupBase]:
+        raise NotImplementedError("This is implemented by the @task decorator!")
+
+
+def task(
+    task_group: Type[TaskGroupBase],
+    required_inputs: Dict[str, ValidInputTypes],
+    output_type: Type[DataBase],
+) -> Callable[[Type[TaskBase]], Type[TaskBase]]:
+    """The decorator for AI Task classes.
+
+    This defines an output data model type for the task, and a minimal set of required inputs
+    that all public models implementing this task must accept.
+
+    As an example, the `caikit.interfaces.nlp.SentimentTask` might look like::
+
+        @task(
+            required_inputs={
+                "raw_document": caikit.interfaces.nlp.RawDocument
+            },
+            output_type=caikit.interfaces.nlp.SentimentPrediction
+        )
+        class SentimentTask(caikit.TaskBase):
+            pass
+
+    and a public model that implements this task might have a .run function that looks like::
+
+        def run(raw_document: caikit.interfaces.nlp.RawDocument,
+                inference_mode: str = "fast",
+                device: caikit.interfaces.common.HardwareEnum) ->
+                    caikit.interfaces.nlp.SentimentPrediction:
+            # impl
+
+    Note the run function may include other arguments beyond the minimal required inputs for
+    the task.
+
+    Args:
+        task_group (Type[TaskGroupBase]): The AI Task Group that this task belongs to
+        required_inputs (Dict[str, ValidInputTypes]): The required parameters that all public
+            models' .run functions must contain. A dictionary of parameter name to parameter
+            type, where the types can be in the set of:
+                - Python primitives
+                - Caikit data models
+                - Iterable containers of the above
+                - Caikit model references (maybe?)
+        output_type (Type[DataBase]): The output type of the task, which all public models'
+            .run functions must return. This must be a caikit data model type.
+
+    Returns:
+        A decorator function for the task class, registering it with caikit's core registry of
+            tasks.
+    """
+    # TODO: type checking on required_inputs
+    if not issubclass(output_type, DataBase):
+        raise TypeError("output_type must be a data model")
+    if not issubclass(task_group, TaskGroupBase):
+        raise TypeError("task_group must be a TaskGroup class")
+
+    for parameter_name, input_type in required_inputs.items():
+        if input_type not in task_group.get_input_type_set():
+            raise TypeError(
+                f"Task parameter {parameter_name} has type {input_type} not in task_group: "
+                f"{task_group.__name__}. Valid types are: {task_group.get_input_type_set()}"
+            )
+
+    def get_required_inputs(_):
```

Comment:
These inline-declared functions should have docstrings so that when a user does `help` on a class wrapped with `@task`, the docs show up nicely